src/plugins/kibana_react/public/code_editor/code_editor.tsx

```diff
@@ -100,118 +104,278 @@ export interface Props {
+    return isMockedComponent ? (ReactMonacoEditor as any)() : ReactMonacoEditor;
+  }, []);
+
+  const _editor = useRef<monaco.editor.IStandaloneCodeEditor | null>(null);
+  const isSuggestionMenuOpen = useRef(false);
+  const editorHint = useRef<HTMLDivElement>(null);
+  const textboxMutationObserver = useRef<MutationObserver | null>(null);
+
+  const [isHintActive, setIsHintActive] = useState(true);
+
+  /* eslint-disable @typescript-eslint/naming-convention */
+  const promptClasses = classNames('kibanaCodeEditor__keyboardHint', {
+    'kibanaCodeEditor__keyboardHint--isInactive': !isHintActive,
+  });
+  /* eslint-enable  @typescript-eslint/naming-convention */
+
+  const _updateDimensions = useCallback(() => {
+    _editor.current?.layout();
+  }, []);
+
+  const startEditing = useCallback(() => {
+    setIsHintActive(false);
+    _editor.current?.focus();
+  }, []);
+
+  const stopEditing = useCallback(() => {
+    setIsHintActive(true);
+  }, []);
+
+  const onKeyDownHint = useCallback(
+    (ev: React.KeyboardEvent) => {
+      if (ev.key === keys.ENTER) {
+        ev.preventDefault();
+        startEditing();
       }
-
-      if (this.props.signatureProvider) {
-        monaco.languages.registerSignatureHelpProvider(
-          this.props.languageId,
-          this.props.signatureProvider
-        );
+    },
+    [startEditing]
+  );
+
+  const onKeydownMonaco = useCallback(
+    (ev: monaco.IKeyboardEvent) => {
+      if (ev.keyCode === monaco.KeyCode.Escape) {
+        // If the autocompletion context menu is open then we want to let ESCAPE close it but
+        // **not** exit out of editing mode.
+        if (!isSuggestionMenuOpen.current) {
+          ev.preventDefault();
+          ev.stopPropagation();
+          stopEditing();
+          editorHint.current?.focus();
+        }
       }
+    },
+    [stopEditing]
+  );
 
-      if (this.props.hoverProvider) {
-        monaco.languages.registerHoverProvider(this.props.languageId, this.props.hoverProvider);
-      }
+  const onBlurMonaco = useCallback(() => {
+    stopEditing();
+  }, [stopEditing]);
 
-      if (this.props.languageConfiguration) {
-        monaco.languages.setLanguageConfiguration(
-          this.props.languageId,
-          this.props.languageConfiguration
-        );
-      }
-    });
-
-    // Register themes
-    monaco.editor.defineTheme('euiColors', this.props.useDarkTheme ? DARK_THEME : LIGHT_THEME);
-    monaco.editor.defineTheme(
-      'euiColorsTransparent',
-      this.props.useDarkTheme ? DARK_THEME_TRANSPARENT : LIGHT_THEME_TRANSPARENT
+  const renderPrompt = useCallback(() => {
+    return (
+      <EuiToolTip
+        display="block"
+        content={
+          <div>
+            <EuiText>
+              <p>
+                <FormattedMessage
+                  id="kibana-react.kibanaCodeEditor.startEditing"
+                  defaultMessage="Press Enter to start editing."
+                />
```

Comment:
I think since we keep this now also for read-only editors we need to change the wording when the editor is `readOnly`. I'd suggest we just use the same wording, that we also use in [EUI](https://github.com/elastic/eui/blob/master/src/components/code_editor/code_editor.tsx#L259-L285).